File: android/java/org/chromium/chrome/browser/allow2/services/Allow2ApiClient.kt
/* Copyright (c) 2024 The Brave Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at https://mozilla.org/MPL/2.0/. */

package org.chromium.chrome.browser.allow2.services

import android.content.Context
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.chromium.chrome.browser.allow2.Allow2Constants
import org.chromium.chrome.browser.allow2.models.*
import org.chromium.chrome.browser.allow2.util.Allow2DeviceInfo
import org.json.JSONArray
import org.json.JSONObject
import java.io.BufferedReader
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL
import javax.net.ssl.HttpsURLConnection

/**
 * HTTP client for Allow2 API communication.
 * Uses HttpURLConnection for compatibility with Chromium's network stack.
 */
class Allow2ApiClient(private val context: Context) {

    private val credentialManager by lazy {
        Allow2CredentialManager.getInstance(context)
    }

    // ==================== Pairing API ====================

    /**
     * Initialize QR code pairing session.
     */
    suspend fun initQrPairing(): ApiResult<QrPairingResponse> = withContext(Dispatchers.IO) {
        try {
            val requestBody = JSONObject().apply {
                put("uuid", Allow2DeviceInfo.getDeviceId(context))
                put("name", Allow2DeviceInfo.getDeviceName(context))
                put("platform", Allow2DeviceInfo.getPlatform())
                put("deviceToken", Allow2DeviceInfo.getDeviceId(context))
            }

            val response = postJson(
                "${Allow2Constants.API_HOST}${Allow2Constants.PAIR_QR_INIT}",
                requestBody
            )

            when (response) {
                is ApiResult.Success -> {
                    val json = response.data
                    val qrResponse = QrPairingResponse(
                        sessionId = json.optString("sessionId"),
                        qrCodeUrl = json.optString("qrCodeUrl"),
                        pin = json.optString("pin", ""),
                        expiresIn = json.optLong("expiresIn", 300)
                    )
                    ApiResult.Success(qrResponse)
                }
                is ApiResult.Error -> response
                is ApiResult.Unauthorized -> response
            }
        } catch (e: Exception) {
            ApiResult.Error("Failed to initialize pairing: ${e.message}")
        }
    }

    /**
     * Poll QR pairing status.
     */
    suspend fun pollQrStatus(sessionId: String): ApiResult<PairingStatusResponse> =
        withContext(Dispatchers.IO) {
            try {
                val response = getJson(
                    "${Allow2Constants.API_HOST}${Allow2Constants.PAIR_QR_STATUS}/$sessionId"
                )

                when (response) {
                    is ApiResult.Success -> {
                        val json = response.data
                        val status = json.optString("status", "pending")

                        val statusResponse = when (status) {
                            "completed" -> {
                                val credentialsJson = json.optJSONObject("credentials")
                                val childrenJson = json.optJSONArray("children")

                                PairingStatusResponse(
                                    status = PairingStatus.COMPLETED,
                                    scanned = true,
                                    credentials = credentialsJson?.let {
                                        Allow2Credentials.fromJson(it)
                                    },
                                    children = childrenJson?.let {
                                        Allow2Child.fromJsonArray(it)
                                    }
                                )
                            }
                            "pending" -> {
                                PairingStatusResponse(
                                    status = PairingStatus.PENDING,
                                    scanned = json.optBoolean("scanned", false),
                                    credentials = null,
                                    children = null
                                )
                            }
                            "expired" -> {
                                PairingStatusResponse(
                                    status = PairingStatus.EXPIRED,
                                    scanned = false,
                                    credentials = null,
                                    children = null
                                )
                            }
                            else -> {
                                PairingStatusResponse(
                                    status = PairingStatus.FAILED,
                                    scanned = false,
                                    credentials = null,
                                    children = null
                                )
                            }
                        }
                        ApiResult.Success(statusResponse)
                    }
                    is ApiResult.Error -> response
                    is ApiResult.Unauthorized -> response
                }
            } catch (e: Exception) {
                ApiResult.Error("Failed to poll status: ${e.message}")
            }
        }

    /**
     * Initialize PIN pairing session.
     */
    suspend fun initPinPairing(): ApiResult<PinPairingResponse> = withContext(Dispatchers.IO) {
        try {
            val requestBody = JSONObject().apply {
                put("uuid", Allow2DeviceInfo.getDeviceId(context))
                put("name", Allow2DeviceInfo.getDeviceName(context))
                put("platform", Allow2DeviceInfo.getPlatform())
                put("deviceToken", Allow2DeviceInfo.getDeviceId(context))
            }

            val response = postJson(
                "${Allow2Constants.API_HOST}${Allow2Constants.PAIR_PIN_INIT}",
                requestBody
            )

            when (response) {
                is ApiResult.Success -> {
                    val json = response.data
                    val pinResponse = PinPairingResponse(
                        sessionId = json.optString("sessionId"),
                        pin = json.optString("pin"),
                        expiresIn = json.optLong("expiresIn", 300)
                    )
                    ApiResult.Success(pinResponse)
                }
                is ApiResult.Error -> response
                is ApiResult.Unauthorized -> response
            }
        } catch (e: Exception) {
            ApiResult.Error("Failed to initialize PIN pairing: ${e.message}")
        }
    }

    // ==================== Service API ====================

    /**
     * Perform usage check for the current child.
     */
    suspend fun check(
        childId: Long,
        activities: List<Allow2Activity> = listOf(Allow2Activity.INTERNET)
    ): ApiResult<Allow2CheckResult> = withContext(Dispatchers.IO) {
        try {
            val credentials = credentialManager.getCredentials()
            if (!credentials.isValid()) {
                return@withContext ApiResult.Error("No valid credentials")
            }

            val activitiesArray = JSONArray().apply {
                activities.forEach { activity ->
                    put(JSONObject().apply {
                        put("id", activity.id)
                        put("log", true)
                    })
                }
            }

            val requestBody = JSONObject().apply {
                put("userId", credentials.userId)
                put("pairId", credentials.pairId)
                put("pairToken", credentials.pairToken)
                put("childId", childId.toString())
                put("activities", activitiesArray)
                put("tz", Allow2DeviceInfo.getTimezone())
            }

            val response = postJson(
                "${Allow2Constants.SERVICE_HOST}${Allow2Constants.SERVICE_CHECK}",
                requestBody
            )

            when (response) {
                is ApiResult.Success -> {
                    val checkResult = Allow2CheckResult.fromJson(response.data)
                    ApiResult.Success(checkResult)
                }
                is ApiResult.Error -> response
                is ApiResult.Unauthorized -> {
                    // 401 means device has been unpaired remotely
                    credentialManager.clearAll()
                    response
                }
            }
        } catch (e: Exception) {
            ApiResult.Error("Check failed: ${e.message}")
        }
    }

    /**
     * Request more time for an activity.
     */
    suspend fun requestMoreTime(
        childId: Long,
        activityId: Int,
        requestedMinutes: Int,
        message: String
    ): ApiResult<RequestResponse> = withContext(Dispatchers.IO) {
        try {
            val credentials = credentialManager.getCredentials()
            if (!credentials.isValid()) {
                return@withContext ApiResult.Error("No valid credentials")
            }

            val requestBody = JSONObject().apply {
                put("userId", credentials.userId)
                put("pairId", credentials.pairId)
                put("pairToken", credentials.pairToken)
                put("childId", childId.toString())
                put("activityId", activityId)
                put("requestedMinutes", requestedMinutes)
                put("message", message)
            }

            val response = postJson(
                "${Allow2Constants.API_HOST}${Allow2Constants.REQUEST_CREATE}",
                requestBody
            )

            when (response) {
                is ApiResult.Success -> {
                    val json = response.data
                    ApiResult.Success(
                        RequestResponse(
                            success = json.optBoolean("success", true),
                            requestId = json.optString("requestId"),
                            message = json.optString("message")
                        )
                    )
                }
                is ApiResult.Error -> response
                is ApiResult.Unauthorized -> {
                    credentialManager.clearAll()
                    response
                }
            }
        } catch (e: Exception) {
            ApiResult.Error("Request failed: ${e.message}")
        }
    }

    // ==================== HTTP Helpers ====================

    private fun postJson(urlString: String, body: JSONObject): ApiResult<JSONObject> {
        var connection: HttpURLConnection? = null
        try {
            val url = URL(urlString)
            connection = (url.openConnection() as HttpsURLConnection).apply {
                requestMethod = "POST"
                setRequestProperty("Content-Type", "application/json")
                setRequestProperty("Accept", "application/json")
                doOutput = true
                connectTimeout = 30000
                readTimeout = 30000
            }

            OutputStreamWriter(connection.outputStream).use { writer ->
                writer.write(body.toString())
                writer.flush()
            }

            val responseCode = connection.responseCode

            return when {
                responseCode == 401 -> ApiResult.Unauthorized()
                responseCode in 200..299 -> {
                    val responseBody = BufferedReader(
                        InputStreamReader(connection.inputStream)
                    ).use { it.readText() }
                    ApiResult.Success(JSONObject(responseBody))
                }
                else -> {
                    val errorBody = try {
                        BufferedReader(
                            InputStreamReader(connection.errorStream)
                        ).use { it.readText() }
                    } catch (e: Exception) {
                        "Unknown error"
                    }
                    ApiResult.Error("HTTP $responseCode: $errorBody")
                }
            }
        } catch (e: Exception) {
            return ApiResult.Error("Network error: ${e.message}")
        } finally {
            connection?.disconnect()
        }
    }

    private fun getJson(urlString: String): ApiResult<JSONObject> {
        var connection: HttpURLConnection? = null
        try {
            val url = URL(urlString)
            connection = (url.openConnection() as HttpsURLConnection).apply {
                requestMethod = "GET"
                setRequestProperty("Accept", "application/json")
                connectTimeout = 30000
                readTimeout = 30000
            }

            val responseCode = connection.responseCode

            return when {
                responseCode == 401 -> ApiResult.Unauthorized()
                responseCode in 200..299 -> {
                    val responseBody = BufferedReader(
                        InputStreamReader(connection.inputStream)
                    ).use { it.readText() }
                    ApiResult.Success(JSONObject(responseBody))
                }
                else -> {
                    val errorBody = try {
                        BufferedReader(
                            InputStreamReader(connection.errorStream)
                        ).use { it.readText() }
                    } catch (e: Exception) {
                        "Unknown error"
                    }
                    ApiResult.Error("HTTP $responseCode: $errorBody")
                }
            }
        } catch (e: Exception) {
            return ApiResult.Error("Network error: ${e.message}")
        } finally {
            connection?.disconnect()
        }
    }

    companion object {
        private const val TAG = "Allow2ApiClient"

        @Volatile
        private var instance: Allow2ApiClient? = null

        fun getInstance(context: Context): Allow2ApiClient {
            return instance ?: synchronized(this) {
                instance ?: Allow2ApiClient(context.applicationContext).also {
                    instance = it
                }
            }
        }
    }
}

// ==================== Response Data Classes ====================

sealed class ApiResult<out T> {
    data class Success<T>(val data: T) : ApiResult<T>()
    data class Error(val message: String) : ApiResult<Nothing>()
    class Unauthorized : ApiResult<Nothing>()
}

data class QrPairingResponse(
    val sessionId: String,
    val qrCodeUrl: String,
    val pin: String,
    val expiresIn: Long
)

data class PinPairingResponse(
    val sessionId: String,
    val pin: String,
    val expiresIn: Long
)

enum class PairingStatus {
    PENDING,
    COMPLETED,
    EXPIRED,
    FAILED
}

data class PairingStatusResponse(
    val status: PairingStatus,
    val scanned: Boolean,
    val credentials: Allow2Credentials?,
    val children: List<Allow2Child>?
)

data class RequestResponse(
    val success: Boolean,
    val requestId: String?,
    val message: String?
)

File: ios/brave-ios/Sources/Brave/Frontend/ParentalFreedom/Allow2APIClient.swift
// Copyright 2024 The Brave Authors. All rights reserved.
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.

import Foundation
import os.log

// MARK: - Allow2 API Client

/// Dedicated network layer for Allow2 API communication
/// Uses URLSession with async/await for all network operations
public final class Allow2APIClient {

  // MARK: - Singleton

  public static let shared = Allow2APIClient()

  // MARK: - Properties

  private lazy var urlSession: URLSession = {
    let config = URLSessionConfiguration.default
    config.timeoutIntervalForRequest = Allow2Constants.Network.requestTimeout
    config.timeoutIntervalForResource = Allow2Constants.Network.resourceTimeout
    config.waitsForConnectivity = true
    return URLSession(configuration: config)
  }()

  private let decoder: JSONDecoder = {
    let decoder = JSONDecoder()
    decoder.keyDecodingStrategy = .convertFromSnakeCase
    return decoder
  }()

  private let encoder: JSONEncoder = {
    let encoder = JSONEncoder()
    encoder.keyEncodingStrategy = .convertToSnakeCase
    return encoder
  }()

  // MARK: - Initialization

  private init() {}

  // MARK: - Pairing API (api.allow2.com)

  /// Initialize QR code pairing session
  /// - Parameters:
  ///   - deviceName: Human-readable name for this device
  ///   - deviceToken: Unique device identifier token
  /// - Returns: Pairing session with QR code URL and session ID
  public func initQRPairing(
    deviceName: String,
    deviceToken: String
  ) async throws -> InitPairingResponse {
    let body = InitPairingRequest(
      uuid: UUID().uuidString,
      name: deviceName,
      platform: "iOS",
      deviceToken: deviceToken
    )

    return try await post(
      url: Allow2Constants.PairingEndpoint.initQR,
      body: body,
      responseType: InitPairingResponse.self
    )
  }

  /// Initialize PIN code pairing session
  /// - Parameters:
  ///   - deviceName: Human-readable name for this device
  ///   - deviceToken: Unique device identifier token
  /// - Returns: Pairing session with PIN code and session ID
  public func initPINPairing(
    deviceName: String,
    deviceToken: String
  ) async throws -> InitPairingResponse {
    let body = InitPairingRequest(
      uuid: UUID().uuidString,
      name: deviceName,
      platform: "iOS",
      deviceToken: deviceToken
    )

    return try await post(
      url: Allow2Constants.PairingEndpoint.initPIN,
      body: body,
      responseType: InitPairingResponse.self
    )
  }

  /// Check pairing status (poll until completed)
  /// - Parameter sessionId: Session ID from init pairing
  /// - Returns: Current pairing status
  public func checkPairingStatus(sessionId: String) async throws -> PairingStatusResponse {
    let url = Allow2Constants.PairingEndpoint.status(sessionId: sessionId)
    return try await get(url: url, responseType: PairingStatusResponse.self)
  }

  /// Cancel active pairing session
  /// - Parameter sessionId: Session ID to cancel
  public func cancelPairing(sessionId: String) async throws {
    let body = CancelPairingRequest(sessionId: sessionId)

    let _ = try await post(
      url: Allow2Constants.PairingEndpoint.cancel,
      body: body,
      responseType: EmptyResponse.self
    )
  }

  // MARK: - Service API (service.allow2.com)

  /// Check usage status and optionally log activity
  /// - Parameters:
  ///   - credentials: Allow2 credentials
  ///   - childId: ID of the child to check
  ///   - activities: Activities to check/log
  ///   - log: Whether to log usage time
  /// - Returns: Check result with activity status
  public func check(
    credentials: Allow2Credentials,
    childId: UInt64,
    activities: [Allow2Activity],
    log: Bool
  ) async throws -> Allow2CheckResult {
    let activityEntries = activities.map { activity -> ActivityEntry in
      return ActivityEntry(id: activity.rawValue, log: log)
    }

    let body = CheckRequest(
      userId: credentials.userId,
      pairId: credentials.pairId,
      pairToken: credentials.pairToken,
      childId: childId,
      activities: activityEntries,
      tz: TimeZone.current.identifier
    )

    return try await post(
      url: Allow2Constants.ServiceEndpoint.check,
      body: body,
      responseType: Allow2CheckResult.self
    )
  }

  // MARK: - Request API (api.allow2.com)

  /// Request more time from parent
  /// - Parameters:
  ///   - credentials: Allow2 credentials
  ///   - childId: ID of the requesting child
  ///   - activity: Activity to request time for
  ///   - duration: Requested duration in minutes
  ///   - message: Optional message to parent
  public func requestMoreTime(
    credentials: Allow2Credentials,
    childId: UInt64,
    activity: Allow2Activity,
    duration: Int,
    message: String?
  ) async throws {
    let body = MoreTimeRequestBody(
      userId: credentials.userId,
      pairId: credentials.pairId,
      pairToken: credentials.pairToken,
      childId: childId,
      activityId: activity.rawValue,
      requestedMinutes: duration,
      message: message
    )

    let _ = try await post(
      url: Allow2Constants.RequestEndpoint.createRequest,
      body: body,
      responseType: EmptyResponse.self
    )
  }

  // MARK: - Generic HTTP Methods

  private func get<T: Decodable>(
    url: URL,
    responseType: T.Type
  ) async throws -> T {
    var request = URLRequest(url: url)
    request.httpMethod = "GET"
    request.setValue(
      Allow2Constants.HTTPHeader.applicationJSON,
      forHTTPHeaderField: Allow2Constants.HTTPHeader.accept
    )
    request.setValue(
      Allow2Constants.HTTPHeader.braveUserAgent,
      forHTTPHeaderField: Allow2Constants.HTTPHeader.userAgent
    )

    return try await execute(request: request, responseType: responseType)
  }

  private func post<T: Encodable, R: Decodable>(
    url: URL,
    body: T,
    responseType: R.Type
  ) async throws -> R {
    var request = URLRequest(url: url)
    request.httpMethod = "POST"
    request.setValue(
      Allow2Constants.HTTPHeader.applicationJSON,
      forHTTPHeaderField: Allow2Constants.HTTPHeader.contentType
    )
    request.setValue(
      Allow2Constants.HTTPHeader.applicationJSON,
      forHTTPHeaderField: Allow2Constants.HTTPHeader.accept
    )
    request.setValue(
      Allow2Constants.HTTPHeader.braveUserAgent,
      forHTTPHeaderField: Allow2Constants.HTTPHeader.userAgent
    )
    request.httpBody = try encoder.encode(body)

    return try await execute(request: request, responseType: responseType)
  }

  private func execute<T: Decodable>(
    request: URLRequest,
    responseType: T.Type
  ) async throws -> T {
    let (data, response) = try await urlSession.data(for: request)

    guard let httpResponse = response as? HTTPURLResponse else {
      Logger.module.error("Allow2 API: Invalid response type")
      throw Allow2APIError.invalidResponse
    }

    // Handle 401 - device unpaired remotely
    if httpResponse.statusCode == 401 {
      Logger.module.warning("Allow2 API: Device unpaired (401)")
      throw Allow2APIError.unauthorized
    }

    // Handle other error status codes
    guard (200...299).contains(httpResponse.statusCode) else {
      Logger.module.error("Allow2 API: Server error \(httpResponse.statusCode)")
      throw Allow2APIError.serverError(httpResponse.statusCode)
    }

    // Handle empty response
    if data.isEmpty, T.self == EmptyResponse.self {
      return EmptyResponse() as! T
    }

    do {
      return try decoder.decode(T.self, from: data)
    } catch {
      Logger.module.error("Allow2 API: Decoding failed: \(error.localizedDescription)")
      throw Allow2APIError.decodingFailed(error)
    }
  }
}

// MARK: - API Error Types

public enum Allow2APIError: Error, LocalizedError {
  case invalidResponse
  case unauthorized
  case serverError(Int)
  case decodingFailed(Error)
  case networkError(Error)

  public var errorDescription: String? {
    switch self {
    case .invalidResponse:
      return "Invalid response from server"
    case .unauthorized:
      return "Device is no longer authorized"
    case .serverError(let code):
      return "Server error: \(code)"
    case .decodingFailed(let error):
      return "Failed to parse response: \(error.localizedDescription)"
    case .networkError(let error):
      return "Network error: \(error.localizedDescription)"
    }
  }
}

// MARK: - Request Models

struct InitPairingRequest: Encodable {
  let uuid: String
  let name: String
  let platform: String
  let deviceToken: String
}

struct CancelPairingRequest: Encodable {
  let sessionId: String
}

struct CheckRequest: Encodable {
  let userId: String
  let pairId: String
  let pairToken: String
  let childId: UInt64
  let activities: [ActivityEntry]
  let tz: String
}

struct ActivityEntry: Encodable {
  let id: Int
  let log: Bool
}

struct MoreTimeRequestBody: Encodable {
  let userId: String
  let pairId: String
  let pairToken: String
  let childId: UInt64
  let activityId: Int
  let requestedMinutes: Int
  let message: String?
}

// MARK: - Response Models

public struct InitPairingResponse: Decodable {
  public let sessionId: String
  public let qrCodeUrl: String?
  public let pinCode: String?
  public let expiresIn: Int?
}

public struct PairingStatusResponse: Decodable {
  public let status: String
  public let scanned: Bool?
  public let completed: Bool
  public let success: Bool
  public let error: String?
  public let credentials: PairingCredentialsResponse?
  public let children: [Allow2Child]?

  enum CodingKeys: String, CodingKey {
    case status
    case scanned
    case completed
    case success
    case error
    case credentials
    case children
  }

  public init(from decoder: Decoder) throws {
    let container = try decoder.container(keyedBy: CodingKeys.self)

    status = try container.decodeIfPresent(String.self, forKey: .status) ?? "pending"
    scanned = try container.decodeIfPresent(Bool.self, forKey: .scanned)

    // Handle completion states
    if let completedValue = try? container.decode(Bool.self, forKey: .completed) {
      completed = completedValue
    } else {
      completed = status == "completed"
    }

    if let successValue = try? container.decode(Bool.self, forKey: .success) {
      success = successValue
    } else {
      success = completed
    }

    error = try container.decodeIfPresent(String.self, forKey: .error)
    credentials = try container.decodeIfPresent(PairingCredentialsResponse.self, forKey: .credentials)
    children = try container.decodeIfPresent([Allow2Child].self, forKey: .children)
  }
}

public struct PairingCredentialsResponse: Decodable {
  public let userId: String
  public let pairId: String
}

struct EmptyResponse: Decodable {
  init() {}
}
